docs/state_encryption.md

```diff
@@ -0,0 +1,303 @@
+As mentioned above, users can configure encryption in environment variables, either as HCL or JSON.
+
+```hcl2
+    key_provider "static" "my_key" {
+      key = "this is my encryption key"
+    }
+    method "aes_gcm" "foo" {
+      key_provider = key_provider.static.my_key
+    }
+    statefile {
+      method = method.aes_gcm.foo
+    }
+```
+
+```json
+{
+  "key_provider" : {
+    "static": {
+      "my_key": {
+        "key": "this is my encryption key"
+      }
+    }
+  },
+  "method": {
+    "aes_gcm": {
+      "foo": {
+        "key_provider": "${key_provider.static.my_key}"
+      }
+    }
+  },
+  "statefile": {
+    "method": "${method.aes_gcm.foo}"
+  }
+}
+```
+
+This data can then be provided to tofu by setting the `TF_ENCRYPTION` environment variable:
+
+```bash
+export TF_ENCRYPTION='{"key_provider":{...},"method":{...},"statefile":{...}}'
+```
+
+To ensure that the encryption is enforced and state cannot be stored unencrypted, the user can specify the `enforced` option in the HCL configuration:
+
+```hcl
+terraform {
+  encryption {
+    //...
+    statefile {
+      enforced = true
+    }
+    planfile {
+      enforced = true
+    }
+    backend {
+      enforced = true
+    }
+    terraform_remote_states {
+      default {
+        enforced = true
+      }
+    }
+  }
+}
+```
+
+**Note:** The `enforced` option is also available in the environment configuration and works as intended, but doesn't make much sense because its primary purpose is to guard against environment variable omission.
+
+## Encrypted state format
+
+When `tofu` encrypts a state file, the encrypted state is still a JSON file. Any implementations can distinguish encrypted files by the `encryption` key being present in the JSON structure. However, there may be other keys in the state file depending on the encryption method and type.
+
+## Encrypted plan format
+
+A plan file in OpenTofu is a ZIP file. However, there are no guarantees that the encrypted plan file will also be a valid ZIP file. The implementation is free to choose any format and there is no way to determine if a plan file is encrypted.
+
+**TODO: why not use the ZIP headers to specify this? Could we use ZIP encryption semantics?**
+
+## Implementation aspect
+
+When implementing the encryption tooling, the implementation should be split in two parts: the library and the OpenTofu implementation. The library should rely on the cty types and hcl as a means to specify schema, but should not be otherwise tied to the OpenTofu codebase. This is necessary to enable encryption capabilities for third party tooling that may need to work with state and plan files.
+
+### Library implementation
+
+The encryption library should create an iinterface that other projects and OpenTofu itself can use to encrypt and decrypt state. The library should express its schema needs (e.g. for key provider config) using [cty](https://github.com/zclconf/go-cty) and hcl, but be otherwise independent of the OpenTofu codebase.
```

Comment:
Should the library include the functionality to parse a configuration in HCL and JSON from the string value of an environment variable into cty? This would free users of the library from the (IMHO very complex) interface of go-cty if all they want to do is decrypt some state file.